flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/enumerator/KafkaSourceEnumerator.java

```diff
@@ -0,0 +1,370 @@
+	private final Set<TopicPartition> discoveredPartitions;
+	/** The current assignment by reader id. Only accessed by the coordinator thread. */
+	private final Map<Integer, Set<KafkaPartitionSplit>> readerIdToSplitAssignments;
+	/** The discovered and initialized partition splits that are waiting for owner reader to be ready. */
+	private final Map<Integer, Set<KafkaPartitionSplit>> pendingPartitionSplitAssignment;
+	/** The consumer group id used for this KafkaSource. */
+	private final String consumerGroupId;
+
+	// Lazily instantiated or mutable fields.
+	private KafkaConsumer<byte[], byte[]> consumer;
+	private AdminClient adminClient;
+	private boolean noMoreNewPartitionSplits = false;
+
+	public KafkaSourceEnumerator(
+			KafkaSubscriber subscriber,
+			OffsetsInitializer startingOffsetInitializer,
+			OffsetsInitializer stoppingOffsetInitializer,
+			Properties properties,
+			SplitEnumeratorContext<KafkaPartitionSplit> context) {
+		this(subscriber, startingOffsetInitializer, stoppingOffsetInitializer, properties, context, new HashMap<>());
+	}
+
+	public KafkaSourceEnumerator(
+			KafkaSubscriber subscriber,
+			OffsetsInitializer startingOffsetInitializer,
+			OffsetsInitializer stoppingOffsetInitializer,
+			Properties properties,
+			SplitEnumeratorContext<KafkaPartitionSplit> context,
+			Map<Integer, Set<KafkaPartitionSplit>> currentSplitsAssignments) {
+		this.subscriber = subscriber;
+		this.startingOffsetInitializer = startingOffsetInitializer;
+		this.stoppingOffsetInitializer = stoppingOffsetInitializer;
+		this.properties = properties;
+		this.context = context;
+
+		this.discoveredPartitions = new HashSet<>();
+		this.readerIdToSplitAssignments = new HashMap<>(currentSplitsAssignments);
+		this.readerIdToSplitAssignments.forEach((reader, splits) ->
+				splits.forEach(s -> discoveredPartitions.add(s.getTopicPartition())));
+		this.pendingPartitionSplitAssignment = new HashMap<>();
+		this.partitionDiscoveryIntervalMs = KafkaSourceOptions.getOption(
+				properties,
+				KafkaSourceOptions.PARTITION_DISCOVERY_INTERVAL_MS,
+				Long::parseLong);
+		this.consumerGroupId = properties.getProperty(ConsumerConfig.GROUP_ID_CONFIG);
+	}
+
+	@Override
+	public void start() {
+		consumer = getKafkaConsumer();
+		adminClient = getKafkaAdminClient();
+		if (partitionDiscoveryIntervalMs > 0) {
+			LOG.info("Starting the KafkaSourceEnumerator for consumer group {} " +
+				"with partition discovery interval of {} ms.", consumerGroupId, partitionDiscoveryIntervalMs);
+			context.callAsync(
+					this::discoverAndInitializePartitionSplit,
+					this::handlePartitionSplitChanges,
+					0,
+					partitionDiscoveryIntervalMs);
+		} else {
+			LOG.info("Starting the KafkaSourceEnumerator for consumer group {} " +
+				"without periodic partition discovery.", consumerGroupId);
+			context.callAsync(
+					this::discoverAndInitializePartitionSplit,
+					this::handlePartitionSplitChanges);
```

Comment:
what would happen if `handlePartitionSplitChanges` (like `context.assignSplits`) failed? Is Kafka source permanently stuck in bad state?

In the other partition discovery enabled case, at least we have a chance to rediscover and reassign.

Reply:
never mind. I guess `addReader` will take care of the above scenario.